### PR TITLE
Fix pulls for gem/me0 in MuonSelectors (and simplify it a bit)

### PR DESCRIPTION
**Pull request by perrotta**
#### PR description:

The relevant part of this PR is the following **fix** for the computation of the pulls in the chamber matching for both GEM and ME0 :
Immagine
(see lines 599-600 and lines 613-614 of DataFormats/MuonReco/src/MuonSelectors.cc )
The pulls were previously computed by normalizing on the square root of the sum of the errors, instead than on the quadratic sum of the same errors.

@ArnabPurohit @trocino could you please check and confirm that such a fix is actually needed? (The alternative is that some weird definition for those pulls was preferred on purpose instead...)

While applying that fix, I profited for integrating also some code simplification meant to improve the performance of the same MuonSelectors:
- Avoid unneeded, and sometimes repeated, computations
- Avoid converting `float` inputs into `double,` and then convert back again to `float` 
- Replace `TMath` functions with `std` ones
- Replace `fabs` with `std::abs`

Further optimizations are still possible:
- Move fully to `float`, also in the methods and their arguments
- Store squares of errors in the muon matches instead of the errors themselves, so that one can avoid repetedly computing square roots and squares when produced and accessed

Regressions are expected:
- whenever the pulls of the matches for GEM and ME0 are used
- probably some very minor additional ones due to possible different rounding approximations because of the other performance related updates

#### PR validation:

It builds, and a few matrix tests succeed.
I didn't try to validate the output by comparing with the baseline
I will try to evaluate possible changes (hopefully an improvement) in cpu performance, and I will update this description with it

### CODE REVIEW

**Comment by perrotta** [comment]
please test

**Comment by perrotta** [comment]
please abort

**Comment by perrotta** [comment]
enable profiling

**Comment by perrotta** [comment]
please test

**Comment by perrotta** [comment]
please test

**Comment by perrotta** [comment]
please test

**Comment by slava77** [comment]
> Reco comparison results: 5832 differences found in the comparisons

there are some pretty large differences in e.g. some MVA values even for Run2 data e.g. 136.88811.
Screen Shot 2020-09-13 at 8 06 13 AM

 This seems to suggest that more than just ME0/GEM were changed.

**Comment by perrotta** [comment]
Thank yo @slava77 , maybe some bug in coding: I will check

**Comment by perrotta** [comment]
please test

**Comment by perrotta** [comment]
There are still differences, and certainly something more to get fixed.
I will keep investigating...

**Comment by perrotta** [comment]
please test

**Comment by perrotta** [comment]
I've measured the CPU time spent in muon ID before and after this PR and, contrarily to my (naive?) expectations I can't notice any significant improvement with this PR. Probably this muon selector did not contribute to a relevant enough fraction of the overall muon ID processing time to make any improvement in it relevant to the total time.
While scrolling through the code I noticed a couple of more places in which some further simplification can be still applied. Since I am on it I think I will update also there, and then I'll let this PR to the evaluation of the muon POG reco contacts, that should certify that the way "pulls" were defined  for in the MuonSelector for GEM and ME0 segment matching were actually wrong, or just wanted as such instead.

File: DataFormats/MuonReco/src/MuonSelectors.cc [hunk 1]
```diff
@@ -179,96 +179,82 @@ float muon::segmentCompatibility(const reco::Muon& muon, reco::Muon::Arbitration
...
+          station_weight[i - 1] = 0.f;
         }
       } else {  // always fully penalize tracks with no matching segment, whether the segment is close to the boundary or not.
         if (station_has_segmentmatch[i - 1] <= 0)
-          station_weight[i - 1] = 0.;
+          station_weight[i - 1] = 0.f;
       }
 
       // if track has matching segment, but the matching is not high quality, penalize
-      if (station_has_segmentmatch[i - 1] > 0 && 42 == 42) {
```
Comment (perrotta) [inline DataFormats/MuonReco/src/MuonSelectors.cc:195]: I'm curious: what was the original purpose of the `&& 42 == 42` part in this condition?
Maybe @drkovalskyi or @slava77 can remember?

**Comment by perrotta** [comment]
please test

File: DataFormats/MuonReco/src/MuonSelectors.cc (hunk 1 above)
Comment (drkovalskyi) [reply DataFormats/MuonReco/src/MuonSelectors.cc:195]: I'm not familiar with this part of the code

Comment (drkovalskyi) [reply DataFormats/MuonReco/src/MuonSelectors.cc:195]: I'm not familiar with this part of the code

**Comment by slava77** [comment]
> Since I am on it I think I will update also there, and then I'll let this PR to the evaluation of the muon POG reco contacts, that should certify that the way "pulls" were defined for in the MuonSelector for GEM and ME0 segment matching were actually wrong, or just wanted as such instead.

It may be more practical to make the technical changes separately, especially if they can lead to small numerical differences.

File: DataFormats/MuonReco/src/MuonSelectors.cc (hunk 1 above)
Comment (slava77) [reply DataFormats/MuonReco/src/MuonSelectors.cc:195]: just a guess,
this looks like a leftover of a pattern `&& true` vs `&& false` from a development stage of a large block of code.

**Comment by trocino** [comment]
> I've measured the CPU time spent in muon ID before and after this PR and, contrarily to my (naive?) expectations I can't notice any significant improvement with this PR. Probably this muon selector did not contribute to a relevant enough fraction of the overall muon ID processing time to make any improvement in it relevant to the total time.
> While scrolling through the code I noticed a couple of more places in which some further simplification can be still applied. Since I am on it I think I will update also there, and then I'll let this PR to the evaluation of the muon POG reco contacts, that should certify that the way "pulls" were defined for in the MuonSelector for GEM and ME0 segment matching were actually wrong, or just wanted as such instead.

Thanks for catching the wrong pull calculation, Andrea. I don't know if there was any reason to define the "pulls" like that — in which case they should probably be called something else. I doubt it though, to me it simply looks like a mistake. Maybe @jshlee and @watson-ij know whether it was intentional or not. Anyway, the performance of these GEM- and ME0-muon selectors will probably be affected significantly by this change, so it's better to inform anyone that uses them for upgrade studies (e.g. in UPSG or GEM POG).
About the changes in CPU time for muon ID, for my information, what sequences are you considering exactly, and what samples are you testing them on? In the standard muon reconstruction sequence, I think muon selectors are used to fill the selector maps for every muon. So the functions you changed (segmentCompatibility and isGoodMuon) should be called multiple times for each muon.

**Comment by jshlee** [comment]
@trocino - I had a look and it does look like the pulls were implemented wrong (#18387). 
We have been meaning to update these selections in the future, once we get ME0 integrated into GEM...

**Comment by perrotta** [comment]
@trocino @jshlee thank you for your checks and your feedback

I think we can go ahead with merging the fix, then. Of course its effect must be evaluated, at least on relvals. However, in all cases these pulls are used to cut on a given number of sigmas, independently on the muon subdet. As such, I don't expect that a tune of the cut parameter has to be applied now because of the fix. Such a cut can be optimized when the new selections @jshlee is talking about will get designed. Do you agree? 

About the (lack of) cpu improvement: I compared timings for some TTbar PU sample, and I did not notice relevant differences overall and, in particular, for the MuonIdProducer module. The same can be observed with the [profiles run in the automatic tests of this PR](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-c2f0e6/9341/profiling/23434.21/RES_CPU_step3.txt) when compared with the [profiles produced with another PR](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-c2f0e6/9309/profiling/23434.21/RES_CPU_step3.txt): if one takes into account the possible fluctuations, I don't see relevant differences in the MuonId modules between them,

When I introduced some bug by mistake I could notice that this MuonSelector is actually run for every muon, as Daniele says. My impression now is that it is not this selector which dominates over the MuonIdProducer time. I have some other simplification in mind for the muonIdProducer, which I will try to test as soon as this PR gets merged,.

**Comment by perrotta** [comment]
+1
- Fixed the pull definition for GEM/ME0 segment matching inside the muon selector: although no effect is visible in the test outputs. it is expected that some will appear in larger stat validations for Run3 and Phase2 workflows
- This PR also simplifies in several parts the flow in the MuonSelectors: also here no effect is visible in the test outputs, but some minor bin flip could become visible for larger stats because of possible different rounding approximations in the new code

**Comment by santocch** [comment]
+1

**Comment by qliphy** [comment]
+1

**Pull request merged**